src/libsyntax/feature_gate.rs

```diff
@@ -1755,6 +1758,15 @@ impl<'a> Visitor<'a> for PostExpansionVisitor<'a> {
             ast::ExprKind::Async(..) => {
                 gate_feature_post!(&self, async_await, e.span, "async blocks are unstable");
             }
+            ast::ExprKind::Call(ref callee, _) => {
+                if let ast::ExprKind::Path(_, ref p) = callee.node {
```

Comment:
Unit struct constructors are not feature gated.

A better place to place the feature gate is probably `lowering.rs`.
Every time we construct `hir::Path {` with a single `Self` segment and `Def::StructCtor` definition we can report a feature error.

Reply:
Thanks. Fixed.